test/integration/cdi/example-driver/controller/controller.go

```diff
@@ -0,0 +1,728 @@
+	if err != nil {
+		return err
+	}
+	if pod.DeletionTimestamp != nil {
+		logger.V(5).Info("Pod marked for deletion")
+		return nil
+	}
+
+	// Still the owner?
+	if !metav1.IsControlledBy(podScheduling, pod) {
+		// Must be obsolete object, do nothing for it.
+		logger.V(5).Info("Pod not owner, PodScheduling is obsolete")
+		return nil
+	}
+
+	// Find all pending claims that are owned by us. We bail out if any of the pre-requisites
+	// for pod scheduling (claims exist, classes exist, parameters exist) are not met.
+	// The scheduler will do the same, except for checking parameters, so usually
+	// everything should be ready once the PodScheduling object exists.
+	var claims []*ClaimAllocation
+	for _, podClaim := range pod.Spec.ResourceClaims {
+		delayed, err := ctrl.checkPodClaim(ctx, pod, podClaim)
+		if err != nil {
+			return fmt.Errorf("pod claim %s: %v", podClaim.Name, err)
+		}
+		if delayed == nil {
+			// Nothing to do for it. This can change, so keep checking.
+			continue
+		}
+		claims = append(claims, delayed)
+	}
+	if len(claims) == 0 {
+		logger.V(5).Info("Found no pending pod claims")
+		return errPeriodic
+	}
+
+	// Check current resource availability *before* triggering the
+	// allocations. If we find that any of the claims cannot be allocated
+	// for the selected node, we don't need to try for the others either
+	// and shouldn't, because those allocations might have to be undone to
+	// pick a better node. If we don't need to allocate now, then we'll
+	// simply report back the gather information.
+	if len(podScheduling.Spec.PotentialNodes) > 0 {
+		if err := ctrl.driver.UnsuitableNodes(ctx, pod, claims, podScheduling.Spec.PotentialNodes); err != nil {
+			return fmt.Errorf("checking potential nodes: %v", err)
+		}
+	}
+	selectedNode := podScheduling.Spec.SelectedNode
+	logger.V(5).Info("pending pod claims", "claims", claims, "selectedNode", selectedNode)
+	if selectedNode != "" {
+		unsuitable := false
+		for _, delayed := range claims {
+			if hasString(delayed.UnsuitableNodes, selectedNode) {
+				unsuitable = true
+				break
+			}
+		}
+
+		if unsuitable {
+			logger.V(2).Info("skipping allocation for unsuitable selected node", "node", selectedNode)
+		} else {
+			logger.V(2).Info("allocation for selected node", "node", selectedNode)
+			selectedUser := &corev1.ResourceClaimUserReference{
+				Version:  "v1",
+				Resource: "pods",
+				Name:     pod.Name,
+				UID:      pod.UID,
+			}
+			for _, delayed := range claims {
+				if err := ctrl.allocateClaim(ctx, delayed.Claim, delayed.ClaimParameters, delayed.Class, delayed.ClassParameters, selectedNode, selectedUser); err != nil {
+					return fmt.Errorf("allocation of pod claim %s failed: %v", delayed.PodClaimName, err)
+				}
+			}
+		}
+	}
+
+	// Now update unsuitable nodes. This is useful information for the scheduler even if
+	// we managed to allocate because we might have to undo that.
```

Comment:
I changed my mind on this one: updating PodScheduling is not useful in most cases, so it should get skipped here after a successful allocation. Instead the code which deallocates should immediately trigger the updating of any pending PodScheduling object.